napari/layers/tracks/tracks.py

```diff
@@ -0,0 +1,520 @@
+        """Sets the view given the indices to slice with."""
+
+        # if the displayed dims have changed, update the shader data
+        if self.dims.displayed != self._current_displayed_dims:
+            # store the new dims
+            self._current_displayed_dims = self.dims.displayed
+            # fire the events to update the shaders
+            self.events.rebuild_tracks()
+            self.events.rebuild_graph()
+
+        return
+
+    def _get_value(self) -> int:
+        """ use a kd-tree to lookup the ID of the nearest tree """
+        coords = np.array(self.coordinates)
+        return self._manager.get_value(coords)
+
+    def _update_thumbnail(self):
+        """Update thumbnail with current points and colors."""
+        colormapped = np.zeros(self._thumbnail_shape)
+        colormapped[..., 3] = 1
+        # return
+        if self._view_data is not None and self.track_colors is not None:
+            de = self._extent_data
+            min_vals = [de[0, i] for i in self.dims.displayed]
+            shape = np.ceil(
+                [de[1, i] - de[0, i] + 1 for i in self.dims.displayed]
+            ).astype(int)
+            zoom_factor = np.divide(
+                self._thumbnail_shape[:2], shape[-2:]
+            ).min()
+            if len(self._view_data) > self._max_tracks_thumbnail:
+                thumbnail_indices = np.random.randint(
+                    0, len(self._view_data), self._max_tracks_thumbnail
+                )
+                points = self._view_data[thumbnail_indices]
+            else:
+                points = self._view_data
+                thumbnail_indices = range(len(self._view_data))
+
+            # get the track coords here
+            coords = np.floor(
+                (points[:, :2] - min_vals[1:] + 0.5) * zoom_factor
+            ).astype(int)
+            coords = np.clip(
+                coords, 0, np.subtract(self._thumbnail_shape[:2], 1)
+            )
+
+            # modulate track colors as per colormap/current_time
+            colors = self.track_colors[thumbnail_indices]
+            times = self.track_times[thumbnail_indices]
+            alpha = (self.current_time - times) / self.tail_length
+            alpha[times > self.current_time] = 1.0
+            colors[:, -1] = np.clip(1.0 - alpha, 0.0, 1.0)
+            colormapped[coords[:, 1], coords[:, 0]] = colors
+
+        colormapped[..., 3] *= self.opacity
+        self.thumbnail = colormapped
+
+    @property
+    def _view_data(self):
+        """ return a view of the data """
+        return self._pad_display_data(self._manager.track_vertices)
+
+    @property
+    def _view_graph(self):
+        """ return a view of the graph """
+        return self._pad_display_data(self._manager.graph_vertices)
+
+    def _pad_display_data(self, vertices):
+        """ pad display data when moving between 2d and 3d
+
+        NOTES:
+            2d data is transposed yx
+            3d data is zyxt
+
+        """
```

Comment:
I'm suuuuper confused about this function... Could you elaborate on it in the docstring? Is this converting from numpy coordinates to VisPy? What's the padding for? You make 2D data flat 3D? And why is that necessary?

Reply:
It's mostly for the case of 2D+t data. In a standard 2D view, the time dimension is projected. However, switching to a 3D view, time becomes a new spatial dimension, and that switch causes an issue with the VisPy rendering. So rendering 2D data as flat 3D stops this from happening when switching between 2D and 3D views of the 2D+t data.